### PR TITLE
add support for wvfpexp in the input layer

### PR DESCRIPTION
**Pull request by totto82**

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

**Comment by blattms** [comment]
You missed adding the new source files to git it seems.

File: tests/parser/ScheduleTests.cpp
Comment (bska) [inline tests/parser/ScheduleTests.cpp:5075]: Tiny nit: These lines (`DIMENS` &c) should really start at the left margin.  Otherwise, it'd be as if the input file were written with leading spaces as
```
        DIMENS
          10 10 10 /
        -- ...
```
instead of typical
```
DIMENS
  10 10 10 /
-- ...
```
The opm-common input parser currently happens to be able to read these such files, but may or may not keep that ability in the future.  I think it's better to stick strictly to the expected format of no leading whitespace characters on keyword lines for our regression tests.

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp
Comment (bska) [inline src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2025]: This does not appear to handle the case of the `WVFPEXP` keyword occurring in an `ACTIONX` block.  I'm okay with postponing that situation until the need arises.

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp
Comment (bska) [inline src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2035]: Tiny optimisation here: You should write this as
```C++
if (well.updateWVFPEXP(std::move(wvfpexp)))
```
instead since we're not going to need the shared `wvfpexp` object after this statement.

File: src/opm/input/eclipse/Schedule/Well/Well.cpp
Comment (bska) [inline src/opm/input/eclipse/Schedule/Well/Well.cpp:519]: Another optimisation opportunity here.  Please write this as
```C++
this->wvfpexp = std::move(wvfpexp_arg);
```
instead.  That way we avoid needlessly incrementing the reference count associated to the `shared_ptr` object.

**Review by bska — COMMENTED** [review]
Looks good for the most part.  There are a couple of needless inefficiencies in handling `shared_ptr<>` objects and we should stick to strict ECLIPSE input format in our unit tests, but otherwise I'm prepared to merge this once the build check passes.

**Comment by totto82** [comment]
Now with the new files.

**Comment by totto82** [comment]
jenkins build this please

File: src/opm/input/eclipse/Schedule/Well/WVFPEXP.cpp
Comment (bska) [inline src/opm/input/eclipse/Schedule/Well/WVFPEXP.cpp:43]: Any reason to not write this in a more canonical form as
```C++
return (m_extrapolate == other.m_extrapolate)
    && (m_shut == other.m_shut)
    && (m_prevent == other.m_prevent);
```

File: opm/input/eclipse/Schedule/Well/WVFPEXP.hpp
Comment (bska) [inline opm/input/eclipse/Schedule/Well/WVFPEXP.hpp:53]: For what it's worth if we're going to create fully compatible restart files we need to be able to distinguish `"YES1"` from `"YES2"` here.  Representing this as a single `bool` blurs that distinction.  Maybe we can represent this as a strong `enum` instead
```C++
enum class Prevent : unsigned char { No, First, Every };
Prevent m_prevent { Prevent::No };
```
and then implement `WVFPEXP::prevent() const` as
```
bool WVPEXP::prevent() const
{
    return m_prevent != Prevent::No;
}
```

**Review by bska — COMMENTED** [review]
Thanks for adding the new structures.  Just a couple of nits there.

File: src/opm/input/eclipse/Schedule/Well/Well.cpp
Comment (bska) [inline src/opm/input/eclipse/Schedule/Well/Well.cpp:370]: Same here; please ensure that we don't reorder member initialisation.

**Review by bska — CHANGES_REQUESTED** [review]
We need to fix the member initialisation order before we merge this.

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp
Comment (totto82) [reply src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2025]: Do we have a "not supported in ACTIONX block" list

**Comment by totto82** [comment]
@bska Thanks for the review. I changed the code accordingly.

**Comment by totto82** [comment]
jenkins build this please

File: src/opm/input/eclipse/Schedule/Well/WVFPEXP.cpp
Comment (bska) [inline src/opm/input/eclipse/Schedule/Well/WVFPEXP.cpp:54]: For safety reasons we should probably assign `Prevent::No` if neither of these conditions hit, e.g.,
```C++
if (prevent_thp == "YES1")
    m_prevent = Prevent::First;
else if (prevent_thp == "YES2")
    m_prevent = Prevent::Every;
else
    m_prevent = Prevent::No;
```

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  With one minor exception this looks good to me now.  I'll run a build check and merge once you've addressed the sole remaining issue.

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp
Comment (bska) [reply src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2025]: > Do we have a "not supported in ACTIONX block" list

I dont' think so, but that would be a good thing to have.

**Comment by OPMUSER** [comment]
@totto82 

> Do we have a "not supported in ACTIONX block" list

It is the opposite you have to add the keyword to the ACTIONX list in order to enable additional SCHEDULE keywords in the ACTIONX block, one needs to add the keyword to:

```
bool ActionX::valid_keyword(const std::string& keyword) {
    static std::unordered_set<std::string> actionx_allowed_list = {
        "COMPLUMP",
        "EXIT",
        "GCONINJE", "GCONPROD", "GCONSUMP", "GLIFTOPT", "GRUPNET", "GRUPTARG", "GRUPTREE", "GSATINJE", "GSATPROD",
        "UDQ",
        "WCONINJE", "WCONPROD", "WECON", "WEFAC", "WELSPECS","WELOPEN", "WELPI", "WGRUPCON", "WPIMULT", "WTEST", "TEST"
    };
    return (actionx_allowed_list.find(keyword) != actionx_allowed_list.end());
}

```
here:

  /home/david/OPM/GitHub/opm-common/src/opm/parser/eclipse/EclipseState/Schedule/Action/ActionX.cpp

**Comment by totto82** [comment]
jenkins build this please

**Comment by bska** [comment]
Build check is green.  I'll merge into master now.  Thanks a lot.

**Pull request merged**
